project/VersionUtil.scala

```diff
@@ -85,12 +86,27 @@ object VersionUtil {
     val date = executeTool("get-scala-commit-date")
     val sha = executeTool("get-scala-commit-sha").substring(0, 7)
 
+    val Patch = """\d+\.\d+\.(\d+)""".r
+    def cross = base match {
+      case Patch(p) if p.toInt > 0 => "bin"
+      case _ => "pre"
+    }
+
     val (canonicalV, mavenSuffix, osgiV, release) = suffix match {
-      case "SNAPSHOT"     => (s"$base-$date-$sha",   s"-SNAPSHOT",      s"$base.v$date-$sha",         false)
-      case "SHA-SNAPSHOT" => (s"$base-$date-$sha",   s"-$sha-SNAPSHOT", s"$base.v$date-$sha",         false)
-      case "SHA-NIGHTLY"  => (s"$base-$sha-nightly", s"-$sha-nightly",  s"$base.v$date-NIGHTLY-$sha", true)
-      case ""             => (s"$base",              "",                s"$base.v$date-VFINAL-$sha",  true)
-      case suffix         => (s"$base-$suffix",      s"-$suffix",       s"$base.v$date-$suffix-$sha", true)
+      // TODO: should this remain `-SNAPSHOT`? does `-SNAPSHOT` have a special meaning for maven / sbt? does `bin-SNAPSHOT` affect this?
+      // TODO: include `-bin` / `-pre` in the canonical version? in the osgi version?
```

Comment:
@dwijnand thanks. Just a detail: IUC, `2.12.1-SNAPSHOT` would also get the short binary version `2.12`, it fits the `NonReleaseV` pattern. However, `2.12.1-sha-nightly` would get the long binary version, as the second `-` doesn't match `\w+`.

Anyway, as you see in this PR, we don't want to rely on this detail, but instead use versions like `2.12.1-bin-.....` and `2.13.0-pre-...`.